### ISSUE TITLE
Problems with Nested ComplexTypes in WSDL

### ISSUE THREAD

**Issue by dotnetdan** [pysimplesoap/pysimplesoap#42]
There are issues interpreting XML schema definitions with nested  complex types.  It looks like there are multiple different symptoms of this issue, but I am providing a repro of a case where the service wsdl defines a property as an array, yet the PySimpleSoap client sees it as a single object.

I've been able to reproduce this using .NET WCF services, defined as such:

![classdiagram1](https://cloud.githubusercontent.com/assets/500965/6508267/d3c72fa2-c325-11e4-8ea4-45aa3eca6ad1.jpg)

The relevant schema from the wsdl looks like this:

``` xml
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:tns="http://schemas.datacontract.org/2004/07/ComplexWsdl" elementFormDefault="qualified" targetNamespace="http://schemas.datacontract.org/2004/07/ComplexWsdl">
    <xs:complexType name="IcedCoffee">
        <xs:complexContent mixed="false">
            <xs:extension base="tns:Coffee">
                <xs:sequence>
                    <xs:element minOccurs="0" name="IceCubes" type="xs:int" />
                </xs:sequence>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>
    <xs:element name="IcedCoffee" nillable="true" type="tns:IcedCoffee" />
    <xs:complexType name="Coffee">
        <xs:complexContent mixed="false">
            <xs:extension base="tns:Beverage">
                <xs:sequence>
                    <xs:element minOccurs="0" name="Beans" nillable="true" type="tns:ArrayOfCoffeeBean" />
                </xs:sequence>
            </xs:extension>
        </xs:complexContent>
    </xs:complexType>
    <xs:element name="Coffee" nillable="true" type="tns:Coffee" />
    <xs:complexType name="Beverage">
        <xs:sequence>
            <xs:element minOccurs="0" name="Ingredients" nillable="true" type="tns:ArrayOfIngredient" />
            <xs:element minOccurs="0" name="Name" nillable="true" type="xs:string" />
        </xs:sequence>
    </xs:complexType>
    <xs:element name="Beverage" nillable="true" type="tns:Beverage" />
    <xs:complexType name="ArrayOfIngredient">
        <xs:sequence>
            <xs:element minOccurs="0" maxOccurs="unbounded" name="Ingredient" nillable="true" type="tns:Ingredient" />
        </xs:sequence>
    </xs:complexType>
    <xs:element name="ArrayOfIngredient" nillable="true" type="tns:ArrayOfIngredient" />
    <xs:complexType name="Ingredient">
        <xs:sequence>
            <xs:element minOccurs="0" name="Name" nillable="true" type="xs:string" />
        </xs:sequence>
    </xs:complexType>
    <xs:element name="Ingredient" nillable="true" type="tns:Ingredient" />
    <xs:complexType name="ArrayOfCoffeeBean">
        <xs:sequence>
            <xs:element minOccurs="0" maxOccurs="unbounded" name="CoffeeBean" nillable="true" type="tns:CoffeeBean" />
        </xs:sequence>
    </xs:complexType>
    <xs:element name="ArrayOfCoffeeBean" nillable="true" type="tns:ArrayOfCoffeeBean" />
    <xs:complexType name="CoffeeBean">
        <xs:sequence>
            <xs:element minOccurs="0" name="Name" nillable="true" type="xs:string" />
            <xs:element minOccurs="0" name="Roast" nillable="true" type="xs:string" />
        </xs:sequence>
    </xs:complexType>
    <xs:element name="CoffeeBean" nillable="true" type="tns:CoffeeBean" />
</xs:schema>
```

The PySimpleSoap client is incorrectly interpreting that the Ingredients property as single object of Ingredient:

```
Port name: BasicHttpBinding_ICoffeeService
Operation Name:  SubmitIcedCoffee
SOAPAction:  http://tempuri.org/ICoffeeService/SubmitIcedCoffee
Input:  {
    "SubmitIcedCoffee": {
        "icedCoffee": {
            "IceCubes": "<type 'int'>",
            "Beans": [
                {
                    "CoffeeBean": {
                        "Name": "<type 'str'>",
                        "Roast": "<type 'str'>"
                    }
                }
            ],
            "Name": "<type 'str'>",
            "Ingredients": {
                "Ingredient": {
                    "Name": "<type 'str'>"
                }
            }
        }
    }
}
Output:  {'SubmitIcedCoffeeResponse': {u'SubmitIcedCoffeeResult': <type 'bool'>}}
```

The code for interpreting the wsdl schema types, is defined in the two recursive routines of helpers.py, postprocess_element and process_element.  It appears that the postprocess_element function is responsible for interpreting extended classes, yet it is failing to fully process all of the complex types when there is an inheritance hierarchy 3 levels deep.  

There are lots of comments mentioning that the recursion needs to be fixed.  If anyone can help or provide any insight on to the issue it would be greatly appreciated!

**Comment by dotnetdan** [comment]
Full WSDL here: http://pastebin.com/3ChLW9vz

**Comment by reingart** [comment]
Hi Dan:

There seems to be an issue with ArrayOfIngredient unbounded not being honored in the Ingredients "alias" (maybe this dialect is not post processing part of the WSDL)

Anyway, I think this should be corrected to avoid "Ingredient" intermediate element:
```
 "Ingredients": [
                {
                    "Name": "<type 'str'>"
                }
            ]
```

If not, it could cause issues with some complex arrays.
You can find more clues looking for the "jetty dialect" in the sources, for example, a recent patch I have to do provisorily:

https://code.google.com/p/pysimplesoap/source/detail?r=52ca02ff27d9d8fb5468b3bb8c529d65e54ea9d1&name=reingart

If you want, I recommend to debug the wsdl parsing stopping when you hit ArrayOfIngredient, let me know if you need more help to analyze this (sorry for not being able help more right now)

**Comment by dotnetdan** [comment]
It's interesting that you mention this.  When I first started using PySimpleSoap, I was very confused by the way lists of objects were handled.  The code sample you wrote above is how I had intuitively expected it to behave.

However, wouldn't this be a breaking change for existing client applications?  Is that something you'd be ok with?

**Comment by reingart** [comment]
Yes, it will break backward compatibility, but also will clean up a lot of quirks dialects
I only applied it for jetty to not break other code, most others webservices use the other "dialect" (.net), for more info you could see the related project:

**Comment by hvdklauw** [comment]
Just ran into the same problem, guess I'm still stuck with suds till this is fixed.

**Comment by dotnetdan** [comment]
@hvdklauw are you having this issue with a .NET WCF server, or a different technology?

**Comment by hvdklauw** [comment]
I actually don't know what server it is, I'll check tomorrow if I can find
out.

**Comment by hvdklauw** [comment]
Ok, my issue is more on the unmarshalling side, I'll create a separate issue

**Comment by dotnetdan** [comment]
@reingart 

I've got this working in a way that the following is working correctly against my WCF service:

```python
result = _client.SubmitIcedCoffee(icedCoffee={
    "Name":"Iced Coffee Dark",
    "IceCubes": 5,
     "Beans":{"CoffeeBean":[{"Name":"Colombian", "Roast":"Medium"}, {"Name":"French", "Roast":"Light"}]},
    "Ingredients": {"Ingredient":[{"Name":"Coffee"}]}
})
```

Note that the complex arrays are 'jetty' style, but intermediate dictionaries are still required, { "CoffeeBean":[...] }  and { "Ingredient":[...] }.

Can you confirm if this is the same for jetty clients?  If not, any pointers on how to prevent the intermediate dictionaries from being required?

**Comment by dotnetdan** [comment]
Can anybody verify if the input in this sample matches the expectation of a jetty service?  I don't have a jetty server to test with.  I'm suspecting that a pysimplesoap client for a jetty service would not have the intermediate "CoffeeBean" and "Ingredient" objects, but I don't have any way to verify.

**Comment by reingart** [comment]
Sorry @dotnetdan , I don't have time to verify this deeply, but  you can take a look in the "jetty" tests cases:

https://github.com/pysimplesoap/pysimplesoap/blob/master/tests/trazamed_test.py

You can also see the real implementation for the "national medical drug traceability program" webservice:

https://github.com/reingart/pyafipws/blob/master/trazamed.py

You can execute it with the following parameters (the last one will return a list of record in the result):
```
python trazamed.py --test
python trazamed.py --testfraccion
python trazamed.py --consulta pruebasws pruebasws
```

There are others webservices client implementations there (for example, wsfev1.py is for .NET, wsaa.py or wsmtx.py are for Java Axis) , if you need I can provide you with test data and the required certificate/private key.

**Comment by soulhunter1987** [comment]
I also ran into this issue with whe following wsdl: http://module.sletat.ru/XmlGate.svc?singlewsdl
When parsing responce PySimpleSoap threw an exception: "TypeError: Tag: Id invalid (type not found)" with the following responce XML:
```xml
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><GetCountriesResponse xmlns="urn:SletatRu:Contracts:Soap11Gate:v1"><GetCountriesResult xmlns:a="urn:SletatRu:DataTypes:CountryCollection:v1" xmlns:i="http://www.w3.org/2001/XMLSchema-instance" xmlns:b="urn:SletatRu:DataTypes:Country:v1"><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">1</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Абхазия</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>2</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">19</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Болгария</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">29</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Вьетнам</Name><b:Alias>NIL</b:Alias><b:Flags>0</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>1</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">35</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Греция</Name><b:Alias>NIL</b:Alias><b:Flags>6</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">39</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Доминикана</Name><b:Alias>NIL</b:Alias><b:Flags>1</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>1</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">40</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Египет</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">44</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Индия</Name><b:Alias>NIL</b:Alias><b:Flags>3</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>1</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">50</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Испания</Name><b:Alias>NIL</b:Alias><b:Flags>6</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">51</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Италия</Name><b:Alias>NIL</b:Alias><b:Flags>6</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">58</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Кипр</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">61</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Куба</Name><b:Alias>NIL</b:Alias><b:Flags>3</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>1</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">72</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Мальдивы</Name><b:Alias>NIL</b:Alias><b:Flags>3</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">77</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Мексика</Name><b:Alias>NIL</b:Alias><b:Flags>3</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>2</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">90</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">ОАЭ</Name><b:Alias>UAE</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">99</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Португалия</Name><b:Alias>NIL</b:Alias><b:Flags>6</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">150</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Россия</Name><b:Alias>Russia</b:Alias><b:Flags>0</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">105</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Сейшелы</Name><b:Alias>SEY</b:Alias><b:Flags>3</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>1</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">113</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Таиланд</Name><b:Alias>Tha</b:Alias><b:Flags>3</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">117</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Тунис</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">119</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Турция</Name><b:Alias>TUR</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">126</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Хорватия</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">127</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Черногория</Name><b:Alias>NIL</b:Alias><b:Flags>2</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country><a:Country><Id xmlns="urn:SletatRu:DataTypes:DbObject:v1">128</Id><Name xmlns="urn:SletatRu:DataTypes:NamedDbObject:v1">Чехия</Name><b:Alias>NIL</b:Alias><b:Flags>4</b:Flags><b:HasTickets>true</b:HasTickets><b:HotelIsNotInStop>true</b:HotelIsNotInStop><b:IsVisa>false</b:IsVisa><b:Rank>0</b:Rank><b:TicketsIncluded>true</b:TicketsIncluded></a:Country></GetCountriesResult></GetCountriesResponse></s:Body></s:Envelope>
``` 
So I made the same change as in #39 - and PySimpleSoap returned OK with just strings instead of some other types.

**Comment by dotnetdan** [comment]
@soulhunter1987 were you able to test is pull request #61 works, instead of disabling strict?

**Comment by soulhunter1987** [comment]
No, I did not. I will try soon and report.

**Comment by soulhunter1987** [comment]
Ok, I've tested it. Generally it works OK. But one field is always returned as None, while there is actually a number of type: <xs:element minOccurs="0" name="BeachLineId" type="xs:unsignedByte"/> It is returned OK as string when in non-strict mode.

**Comment by dotnetdan** [comment]
Thanks Dmitry.  If you can provide a sample repro I will try to address it.

**Comment by dotnetdan** [comment]
@soulhunter1987 Actually, can you confirm if this as an issue specifically related to "unsignedByte"?  Do you have any other unsignedByte fields that work?

**Comment by dotnetdan** [comment]
Should be able to fix your issue in helpers.py, by adding 'unsignedByte' to the REVERSE_TYPE_MAP

```python
REVERSE_TYPE_MAP.update({
    'base64Binary': str,
    'unsignedByte': byte
})
```

**Comment by soulhunter1987** [comment]
Yes, it fixes the issue. Could you, please, make a commit to this branch so that we could use it "oficially" )
Also there are some more interesting types in my WSDL: unsignedInt, unsignedLong, unsignedShort. There are actually no fields, that use these types, but i think they should also be added to REVERSE_TYPE_MAP for further compliance.
And when do you plan to merge this branch into master and make a release?

**Comment by soulhunter1987** [comment]
Also, I think that current behavior is incorrect. It should throw an exception with unknown type instead of returning None in this case, cause other, unexperiences users may even not mension incorrect output data without explicit error returned.

**Comment by dotnetdan** [comment]
@soulhunter1987 Can you create a separate issue for the missing type map?  If you could create a pull request I'd be glad to merge it for you.

For #61, I would like to merge this, but it's a breaking change to existing WCF service clients and I'm not comfortable going forward without some sort of validation from @reingart and the greater community.  Mariano, do you have any thoughts or suggestions on how to proceed from here?

**Comment by soulhunter1987** [comment]
You may make a new major release and that won't be a breking change since major version updates usually asume api breaking changes verification.

**Comment by dotnetdan** [comment]
I've merged pull request #61 to address this issue.  I'd like to get this into a major release as soon as possible, any feedback on this change will be greatly appreciated.
